File: LC/[1589] Maximum Sum Obtained of Any Permutation.cpp
class Solution {
public:
    int maxSumRangeQuery(vector<int>& nums, vector<vector<int>>& requests) {
        int N = nums.size();
        vector<int> events(N+1, 0);

        for(auto request: requests){
            events[request[0]] += 1;
            events[request[1]+1] -=1;
        }

        vector<int> freq(N);
        int pre = 0;
        for(int ii=0; ii<N; ++ii){
            pre = pre + events[ii];
            freq[ii] = pre;
        }

        sort(freq.begin(), freq.end());
        sort(nums.begin(), nums.end());

        long ret = 0;
        long M = 1e9+7;
        for(int ii=0; ii<N; ++ii){
            ret = (ret + (long)freq[ii]*(long)nums[ii]) %M;
        }
        return ret;
    }
};

File: LC/[1589] Maximum Sum Obtained of Any Permutation.py
class Solution:
    def maxSumRangeQuery(self, nums: List[int], requests: List[List[int]]) -> int:
        N=len(nums)
        events = [0]*(N+1) #considering the ending is +1
        # 1. convert interval requests to events
        for (st, ed) in requests:
            events[st] += 1
            events[ed+1]-=1

        freq = [0]*N
        freq[0] = events[0]
        for ii in range(1, N):
            freq[ii] = freq[ii-1]+events[ii]
        
        freq.sort()
        nums.sort()

        M=pow(10, 9) + 7
        ttl=0
        for ii in range(N):
            ttl = (ttl + freq[ii]*nums[ii]%M)%M

        return ttl




"""
* requesting for a range -> range sum
* num of request is frequency on on an index
* the value of this index can produce -> freqXval
* we need to assing highest value to highest freq
* how to calc freq of request -> range request -> sweepline
"""
        
    